sklearn/cluster/_k_means.pyx

```diff
@@ -8,167 +8,207 @@
 #
 # License: BSD 3 clause
 
-from libc.math cimport sqrt
 import numpy as np
-import scipy.sparse as sp
 cimport numpy as np
 cimport cython
 from cython cimport floating
+from libc.math cimport sqrt
+
+
+np.import_array()
 
-from sklearn.utils.sparsefuncs_fast import assign_rows_csr
 
 ctypedef np.float64_t DOUBLE
 ctypedef np.int32_t INT
 
-cdef extern from "cblas.h":
-    double ddot "cblas_ddot"(int N, double *X, int incX, double *Y, int incY)
-    float sdot "cblas_sdot"(int N, float *X, int incX, float *Y, int incY)
 
-np.import_array()
+cpdef floating _inertia_dense(np.ndarray[floating, ndim=2, mode='c'] X,
```

Comment:
Can you confirm my understanding that you're not using a memview here because `floating` is a fused type and `X` might be read-only (https://github.com/scikit-learn/scikit-learn/issues/10624#issuecomment-449407648)?

Would it be acceptable (/useful?) to always convert `X` to float32 or float64 and leverage memory views?

Reply:
This is the reason indeed.

I'm not sure about converting. This means always making a copy of X. It might be bad for large datasets, although `copy_x` is True by default. I don't mind deprecating `copy_x`, especially since not copying X leads to less numerical precision.